### ISSUE TITLE
Link not created if followed by a single colon.

### ISSUE THREAD

**Issue by reylau** [doxygen/doxygen#7165]
I have joined an example.

I try to create a link to an anchor. The `@ref AnchorName` works fine __but__ if `AnchorName` is followed by a single `:`. In that case, Doxygen generates a warning (see "Cases\2 Doxygen log.txt") and the link is not created (see "Cases\3 Output.PNG").

I understand that a double colon has a special meaning, but why a single colon fails?
[2019-08-06 Doxygen Bug xxxx.zip](https://github.com/doxygen/doxygen/files/3472658/2019-08-06.Doxygen.Bug.xxxx.zip)

**Comment by albert-github** [comment]
I looked into the doctokenizer.l` and the "offending" line looks like to be line 1056:
```
<St_Ref>{REFWORD}      { // label to refer to
```
where there are 2 matches: `REFWORD2` and `REFWORD3` defined as (line 411 ... 418 of the same file):
```
REFWORD2_PRE   ("#"|"::")?((({ID}{TEMPLPART}?)|{ANONNS})("."|"#"|"::"|"-"|"/"))*({ID}{TEMPLPART}?(":")?)
REFWORD2       {REFWORD2_PRE}{FUNCARG2}?
REFWORD2_NOCV  {REFWORD2_PRE}("("{FUNCPART}")")?
REFWORD3       ({ID}":")*{ID}":"?
REFWORD4_NOCV  (({SCOPEPRE}*"operator"{OPMASKOP2})|(("::"|"#"){SCOPEPRE}*"operator"{OPMASKOP2}))
REFWORD4       {REFWORD4_NOCV}{CVSPEC}?
REFWORD        {FILEMASK}|{LABELID}|{REFWORD2}|{REFWORD3}|{REFWORD4}
REFWORD_NOCV   {FILEMASK}|{LABELID}|{REFWORD2_NOCV}|{REFWORD3}|{REFWORD4_NOCV}
```

It is unclear for me which cases are meant by the single colon end of `REFWORD2` (actually `REFWORD2_PRE` as  `{FUNCARG2} ` is optional) and `REFWORD3`

@doxygen any insight?

**Comment by doxygen** [comment]
@albert-github the colons are for referring to Objective-C methods. See https://roadfiresoftware.com/2014/06/how-to-read-and-write-method-declarations-in-objective-c/ for more information about how they are structured. In the referenced example you have `driveCar:withPerson:` as a method name.

**Comment by albert-github** [comment]
So when we are not in object C:
- we could always ignore `REFWORD3`
- in case the matches `REFWORD2` and it doesn't contain the `":"` part also ignore it

**Comment by doxygen** [comment]
@albert-github Problem is how to known if "we are in objective C". This is about a reference in a comment block, which does not necessarily have to be inside an objective C file (and `.h` files are also used for Objective-C code, so from the file's extension it is also not clear).

**Comment by albert-github** [comment]
Indeed in the `doctokenizer.l` it is not known which language we are in and with a `.h` file it is even more difficult as this can be used by multiple languages.

**Comment by mosra** [comment]
FYI I'm working around this issue by doing `@link AnchorName @endlink:` instead of `@ref AnchorName:`. It's slightly annoying to type compared to a simple `@ref`, but it always worked reliably in this case.

I was not aware about this ObjC syntax, that explains a lot.

**Comment by reylau** [comment]
@mosra Thank you for this workaround. In my case, I addressed this by adding a blank between the link and `:`.

@albert-github Would it be an idea to have a Doxyfile flag to define if this is Objective-C ? Or / and maybe a special tag to be used in source files?

**Comment by albert-github** [comment]
@reylau 
Some thoughts:
- I think a general extra configuration setting I don't think will work as it will either work for all comments, but probably giving problems in multi language cases or in case a single colon is required in Obj-C. Specifying all files that are exempt from the Obj-C behavior would, I think, also not be feasible as it would be quite illogical.
- the only special tag I can think of is `\:`, maybe a solution (analogous to the already existing `\::`).

**Comment by mosra** [comment]
Another idea: could it try to find a symbol without the trailing `:` if it's not found with? That could work reliably.

I don't think adding any configuration settings or language detection is a good idea, it only adds more combinations and increases the chance of new problems appearing.

**Comment by albert-github** [comment]
@mosra 
Regarding the "could it try to find a symbol without the trailing : if it's not found with"do you mean first check with the, single, colon and when not found try without the single colon?

**Comment by mosra** [comment]
Yes, with `@ref Thing:` it would first look for `Thing:` and if that's not found then `Thing`.

(I assume such code is already in place, because from my usage it sometimes felt like it's backtracking when it didn't find the whole thing ... or maybe it's not?)

**Comment by reylau** [comment]
I agree with your comments on new flag, tag and find that the last suggestion from mosra is an elegant  solution, i.e. search first for `Thing:` then `Thing`. I would be happy with it!
